addon-chrome/ChromePlacesProvider.js

```diff
@@ -0,0 +1,283 @@
+const db = require("addon-chrome/db");
+
+const {BOOKMARK, HISTORY, BLOCKED_URL, LAST_VISIT_TIME, DATE_ADDED, PREV} = require("addon-chrome/constants");
+const {getMetadata} = require("page-metadata-parser");
+
+module.exports = class ChromePlacesProvider {
```

Comment:
Your strategy seems to be the following:
- on start, you copy over bookmarks and history from chrome.history/bookmarks API into a local DB, doing some data transformations in the process
- you subscribe for updates to chrome.history/bookmarks API, and keep updating your local DB
- when data is requested, you query your local DB, do some transformations and then return

So, my question: what is the intention of buffering data in a local DB? If you remove it from the equation, when content (activity stream app) requests data (highlights, history, etc), you will query chrome.history/bookmarks APIs directly, do any necessary work and return.

For history and bookmarks I see no strong need for the extra caching layer.

You do have to deal with metadata about pages (image, description, etc), which you query asynchronously. This can be a simple key/value store based on URL, or domain itself, or a combination of the two, which you query against while fetching data from the chrome APIs. If metadata is present, then merge it into what you're returning. Otherwise, schedule a fetch and then push an update to activity stream app when it resolves. To make this flow nicer, you may consider batching these fetches as well, and return when Promise.all(batchedFetches) is satisfied, for example.